File: IntentFilter/app/src/main/java/com/xfhy/intentfilter/MainActivity.java
package com.xfhy.intentfilter;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.btn_test).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent();
                //intent.setAction("com.xfhy.aa");  //action只要匹配到一个就行

                intent.addCategory("android.intent.category.DEFAULT");   //category可以不用添加
                intent.setType("text/plain");

                //必须先判断一下,毕竟是启动隐式Activity
                if (intent.resolveActivity(getPackageManager()) != null) {
                    startActivity(intent);
                }
            }
        });

    }
}

File: BroadCastReceiverDemo/app/src/main/java/com/xfhy/broadcastreceiverdemo/MainActivity.kt
package com.xfhy.broadcastreceiverdemo

import android.content.Intent
import android.content.IntentFilter
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val dynamicBroadcast = DynamicBroadcast()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        btn_static.setOnClickListener {
            //用Activity或者applicationContext都行
//            val intent = Intent(baseContext, StaticReceiver::class.java)
            val intent = Intent(applicationContext, StaticReceiver::class.java)
            intent.action = "com.xfhy.staticreceiver"
            sendBroadcast(intent)
        }

        btn_dynamic.setOnClickListener {
            val intentFilter = IntentFilter()
            registerReceiver(dynamicBroadcast, intentFilter)

            val intent = Intent(this, DynamicBroadcast::class.java)
            sendBroadcast(intent)
        }

    }

    override fun onStop() {
        super.onStop()
        unregisterReceiver(dynamicBroadcast)
    }

}
